### PR TITLE
give the sys.path to the process_cmdline for dev mode

### PR DESCRIPTION
**Pull request by servoz**
Testing the populse_mia [PR #165](https://github.com/populse/populse_mia/pull/165), it was observed that the populse_mia dev mod was broken by a system call loosing the sys.path defined in the populse_mia main.py. This little change could fix it.

### CODE REVIEW

File: capsul/pipeline/pipeline_workflow.py [hunk 1]
```diff
@@ -360,10 +360,13 @@ def _replace_dict_transfers(rdict, process, itransfers, otransfers):
         use_input_params_file = False
         if process_cmdline[0] == 'capsul_job':
             python_command = os.path.basename(sys.executable)
+            # pass the sys.path for dev mode
             process_cmdline = [
                 'capsul_job', python_command, '-c',
+                'import sys; sys.path = %s; '
                 'from capsul.api import Process; '
-                'Process.run_from_commandline("%s")' % process_cmdline[1]]
+                'Process.run_from_commandline("%s")' % (sys.path,
```
Comment (sapetnioc) [inline capsul/pipeline/pipeline_workflow.py:368]: The syntax is confusing here. By luck, the tuple used in the last `%` operator work on the two remaining `%s` in the string. But between these two `%s` there is another `%s` with its `%` operator. This is too difficult to understand. Either, each string with `%s` should have their `%` operator, or the whole string should be within parenthesis and a single `%` operator used with a 3 values tuple.

**Comment by servoz** [comment]
I acknowledge that I was only thinking of the case of local execution when I proposed this change and, I agree, injecting the sys.path locally valid may create problems in another execution environment. But the fact is that without this change the local execution just doesn't work if mia is run in developer mode (which makes me a little bit nervous because I only use mia in developer mode!). I think that if we don't use the method I propose  we should propose another one ! 
We had already talked about an attribute, a flag or other that would indicate if we are in local or remote mode. In this case it would be enough to have a loop if flag == local then inject the sys.path as I suggest (in the case of a deported calculation we can consider that we will only run in user mode, so no more problem.
Another solution would be to do what you propose. In this case I confess I'm not sure I understood the details well enough to code it myself.

File: capsul/pipeline/pipeline_workflow.py (hunk 1 above)
Comment (servoz) [reply capsul/pipeline/pipeline_workflow.py:368]: I confess I don't understand your point.
I don't regularly use the C style formatting % operator (I generally prefer the format() method) but I wanted to stay in the coding style existing in this module.
Are you saying that the following syntax is not good? 
`'%s like %s' % ('I', 'strawberries.')`

**Comment by denisri** [comment]
What I understand of the changes in this PR is to completely skip the config system when running python, and to hard-code the current python executable and `sys.path` into the executed command. This will work in your situation (local execution and same execution config as the running client), but will not work in all cases, especially in remote execution, and will not allow config customization at all.
As things (python executable and path) were already present in the config system, I understand that your problem is merely a matter of providing a default config for the local computing resource that matches the running client config, am I wrong ?

**Comment by servoz** [comment]
We could put it like that!
The default configuration of the local computer resource would mainly correspond to where we find the populse packages and it should change automatically at the startup between the site package and the directory hosting the scripts being developed.  Better than a long description, this is currently done in the first lines of mia's main.py:

```
# Adding populse_mia path to the sys.path if in developer mode
if not os.path.dirname(os.path.dirname(
        os.path.realpath(__file__))) in sys.path:           # "developer" mode
    DEV_MODE = True
    root_dev_dir = os.path.dirname(
                     os.path.dirname(
                       os.path.dirname(
                         os.path.dirname(os.path.realpath(__file__)))))
    i = 0
    # Adding populse_mia
    print('\n- Mia in "developer" mode')
    mia_dev_dir = os.path.join(root_dev_dir, 'populse_mia', 'python')
    print('  . Using populse_mia package from {} ...'.format(mia_dev_dir))
    sys.path.insert(i, mia_dev_dir)
    del mia_dev_dir
    from populse_mia import info
    print(f"    populse_mia version : {info.__version__}")

    # Adding capsul
    if os.path.isdir(os.path.join(root_dev_dir, 'capsul')):
        i += 1
        capsul_dev_dir = os.path.join(root_dev_dir, 'capsul')
        print('  . Using capsul package from {} ...'.format(capsul_dev_dir))
        sys.path.insert(i, capsul_dev_dir)
        del capsul_dev_dir

    else:
        import capsul
        capsul_dir = os.path.dirname(os.path.dirname(capsul.__file__))
        print('  . Using capsul package from {} ...'.format(capsul_dir))
        del capsul_dir
        del capsul

     # Adding soma_base:
    if os.path.isdir(os.path.join(root_dev_dir, 'soma-base')):
        i += 1
        soma_b_dev_dir = os.path.join(root_dev_dir, 'soma-base', 'python')
        print('  . Using soma package from {} ...'.format(soma_b_dev_dir))
        sys.path.insert(i, soma_b_dev_dir)
        del soma_b_dev_dir

    else:
        import soma
        soma_b_dir = os.path.dirname(os.path.dirname(soma.__file__))
        print('  . Using soma package from {} ...'.format(soma_b_dir))
        del soma_b_dir
        del soma

    # Adding soma_workflow:
    if os.path.isdir(os.path.join(root_dev_dir, 'soma-workflow')):
        i += 1
        soma_w_dev_dir = os.path.join(root_dev_dir, 'soma-workflow', 'python')
        print('  . Using soma_workflow package from {} '
              '...'.format(soma_w_dev_dir))
        sys.path.insert(i, soma_w_dev_dir)
        del soma_w_dev_dir

    else:
        import soma_workflow
        soma_w_dir = os.path.dirname(os.path.dirname(soma_workflow.__file__))
        print('  . Using soma_worflow package from {} ...'.format(soma_w_dir))
        del soma_w_dir
        del soma_workflow

    # Adding populse_db:
    if os.path.isdir(os.path.join(root_dev_dir, 'populse_db')):
        i += 1
        populse_db_dev_dir = os.path.join(root_dev_dir, 'populse_db', 'python')
        print('  . Using populse_db package from {} '
              '...'.format(populse_db_dev_dir))
        sys.path.insert(i, populse_db_dev_dir)
        del populse_db_dev_dir

    else:
        import populse_db
        populse_db_dir = os.path.dirname(os.path.dirname(populse_db.__file__))
        print('  . Using populse_db package from {} ...'.format(populse_db_dir))
        del populse_db_dir
        del populse_db

    # Adding mia_processes:
    if os.path.isdir(os.path.join(root_dev_dir, 'mia_processes')):
        i += 1
        mia_processes_dev_dir = os.path.join(root_dev_dir,
                                             'mia_processes',
                                             'python')
        print('  . Using mia_processes package from {} '
              '...'.format(mia_processes_dev_dir))
        sys.path.insert(i, mia_processes_dev_dir)
        del mia_processes_dev_dir

    else:

        try:
            import mia_processes

        except Exception:
            pass

        else:
            mia_processes_dir = os.path.dirname(
                                        os.path.dirname(mia_processes.__file__))
            print('  . Using mia_processes package from {} '
                  '...'.format(mia_processes_dir))
            del mia_processes_dir
            del mia_processes

    del root_dev_dir
```
So the sys.path changes depending on how we start mia ...

**Comment by denisri** [comment]
Does populse/populse_mia@7f1c811 help ?

**Comment by servoz** [comment]
I currently have a crash at starting time:
```
$ mia

- Mia in "developer" mode
  . Using populse_mia package from /data/Git_Projects/populse_mia/python ...
    populse_mia version : 1.4.0-dev+8c3c24d4
  . Using capsul package from /data/Git_Projects/capsul/ ...
  . Using soma package from /data/Git_Projects/soma-base/python ...
  . Using soma_workflow package from /data/Git_Projects/soma-workflow/python ...
  . Using populse_db package from /data/Git_Projects/populse_db/python ...
  . Using mia_processes package from /data/Git_Projects/mia_processes/python ...

/home/econdami/.populse_mia/configuration.yml has been detected.
Traceback (most recent call last):
  File "main.py", line 1106, in <module>
    main()
  File "main.py", line 661, in main
    pc = c['engine'].setdefault('global', {}).setdefault('capsul.engine.module.python', {})
KeyError: 'engine'



```

**Comment by servoz** [comment]
So thanks to https://github.com/populse/populse_mia/issues/172 it is working with db 2.0 and mia populse_db2_capsulengine and capsul 4PR165onMIA (master not working yet) if the next part is commented in  current populse_db2_capsulengine:
```
    """
    if DEV_MODE and pypath:
        config = Config()
        config.get_capsul_engine()
        c = config.get_capsul_config()
        pc = c['study_config'].setdefault('global', {}).setdefault('capsul.engine.module.python', {})
        if not pc.get('executable'):
            pc['executable'] = sys.executable
        if not pc.get('path'):
            pc['path'] = pypath
        print('changed python conf:', pc)
        config.update_capsul_config()
        config.saveConfig()
    """
```

**Comment by servoz** [comment]
A little bit of introspection:
By using (by uncommenting)

```
    if DEV_MODE and pypath:
        config = Config()
        config.get_capsul_engine()
        c = config.get_capsul_config()
        pc = c['study_config'].setdefault('global', {}).setdefault('capsul.engine.module.python', {})
        if not pc.get('executable'):
            pc['executable'] = sys.executable
        if not pc.get('path'):
            pc['path'] = pypath
        print('changed python conf:', pc)
        config.update_capsul_config()
        config.saveConfig()
```
    
Mia crash at startup:
```
Traceback (most recent call last):
  File "main.py", line 1109, in <module>
    main()
  File "main.py", line 663, in main
    pc = c['engine'].setdefault('global', {}).setdefault('capsul.engine.module.python', {})
KeyError: 'engine'
```

Indeed, the c dictionnary object that is created has no engine key at this point:
`c:  {'engine_modules': ['nipype', 'fsl', 'freesurfer', 'matlab', 'spm', 'fom', 'python'], 'study_config': {'attributes_schema_paths': ['capsul.attributes.completion_engine_factory', 'populse_mia.user_interface.pipeline_manager.process_mia'], 'process_completion': 'mia_completion', 'spm_exec': '/usr/local/SPM/spm12_standalone/run_spm12.sh', 'use_spm': True, 'spm_directory': '/usr/local/SPM/spm12_standalone', 'spm_standalone': True}}
`

**Comment by denisri** [comment]
I have pushed an attempt to fix this (populse/populse_mia@31a83e2376edd07dd1c00cba7bc6b1a), could you tell me if it works better for you ?

**Comment by servoz** [comment]
So now mia no longer crashes at startup, thanks to your last commit ! and it is now possible to launch the smooth brick from the mia_processes library. Cool !
But this is yet true only with the 4PR165onMIA branch of capsul (this PR subject). By switching to the master branch of capsul mia crash with:
r```
unning pipeline...
soma-workflow starting in light mode
Workflow controller initialised

 When the pipeline was launched, the following exception was raised: Error during workflow execution. Status=('finished_regularly', 1, None, None).
The workflow has not been removed from soma_workflow and must be deleted manually. 
Failed jobs: [76]
Aborted/killed jobs: []
============================================
---- failed job info ---
* job: 76: smooth1
* exit status: finished_regularly
* commandline:
'python3' '-c' 'from capsul.api import Process; Process.run_from_commandline("mia_processes.preprocess.spm.spatial_preprocessing.Smooth")'
* exit value: 1
* term signal: None
---- env ----
None
---- configuration ----
{'capsul_engine': {'uses': {'capsul.engine.module.spm': 'any'}}, 'capsul.engine.module.spm': {'config_id': 'spm12-standalone', 'config_environment': 'global', 'directory': '/usr/local/SPM/spm12_standalone', 'version': '12', 'standalone': True}, 'capsul.engine.module.python': {'capsul_engine': {'uses': {'capsul.engine.module.python': 'any'}}}}
---- stdout ----

---- stderr ----
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'capsul'
 ...
```

**Comment by servoz** [comment]
Thanks to latest @denisri's commits on mia package, it seems that this PR is now obsolete. I close it !

**Pull request closed without merging**